File: recDP/06-maxSubsetNoAdjSum/my.py
def maxSubsetSumNoAdjacent(array):
    # Write your code here.

	if len(array) == 0:
		return 0
	return max(array[0] + maxSubsetSumNoAdjacent(array[2:]), maxSubsetSumNoAdjacent(array[1:]))

def maxSubsetSumNoAdjacent(array, memo={}):
    # Write your code here.
    if len(array) in memo:
        return memo[len(array)]
    if len(array) == 0:
        return 0
    memo[len(array)] = max(array[0] + maxSubsetSumNoAdjacent(array[2:], memo), maxSubsetSumNoAdjacent(array[1:], memo))
    return memo[len(array)]

print(maxSubsetSumNoAdjacent([7, 10, 12, 7, 9, 14, 15, 16, 25, 20, 4]) ==  72)

File: old/recDP/08-minNumCoins/my.cpp
#include <vector>
#include <climits>
#include <algorithm>
using namespace std;

int fn(int n, vector<int> denoms) {
  // Write your code here.
	if (n < 0)
		return 10000; // repr of INT_MAX + 1 is gonna be INT_MIN
	if (n == 0)
		return 0;
	if (denoms.size() == 0)
		return 10000;
	vector<int> nv(denoms.begin(), denoms.end() - 1);
	return min(1 + fn(n - denoms[denoms.size() - 1], denoms), fn(n, nv));


}

int minNumberOfCoinsForChange(int n, vector<int> denoms) {

	int ans = fn(n, denoms);
	return ans < 10000 ? ans : -1;
}
